File: backend/core/responses/basePlaylistResponse.py
from typing import Generic, List, Literal, Optional, Sequence, TypeVar, Union
from datetime import datetime

from pydantic import BaseModel, field_serializer

from backend.core.enums.playlistContributorRoleEnum import PlaylistContributorRoleEnum
from backend.core.responses.baseAlbumWithoutSongsResponse import (
    BaseAlbumWithoutSongsResponse,
)
from backend.core.responses.basePlaylistForPlaylistResponse import (
    BasePlaylistForPlaylistResponse,
)
from backend.core.responses.baseSongWithAlbumResponse import BaseSongWithAlbumResponse
from backend.core.responses.baseStationResponse import BaseStationResponse
from backend.core.responses.baseVideoResponse import BaseVideoResponse

T = TypeVar("T")


class PlaylistContributorResponse(BaseModel):
    user_id: int
    role: PlaylistContributorRoleEnum

    @field_serializer("role")
    def serialize_queue_type(self, role: PlaylistContributorRoleEnum) -> str:
        return role.name


class PlaylistResponseItem(BaseModel, Generic[T]):
    item: T
    addedAt: datetime


class BasePlaylistResponse(BaseModel):
    type: Literal["playlist"] = "playlist"
    description: Optional[str]
    provider: str
    publicId: str
    url: str
    providerUrl: str
    name: str
    medias: Sequence[
        Union[
            PlaylistResponseItem[BaseSongWithAlbumResponse],
            PlaylistResponseItem[BaseVideoResponse],
            PlaylistResponseItem[BaseStationResponse],
            PlaylistResponseItem[BasePlaylistForPlaylistResponse],
            PlaylistResponseItem[BaseAlbumWithoutSongsResponse],
        ]
    ]
    contributors: List[PlaylistContributorResponse]
    imageUrl: str
    owner: str

File: frontend/packages/shared/src/dto/basePlaylistResponse.ts
import { z } from "zod";
import { BaseAlbumWithoutSongsResponseSchema } from "./baseAlbumWithoutSongsResponse";
import { BaseAlbumWithSongsResponseSchema } from "./baseAlbumWithSongsResponse";
import { BasePlaylistForPlaylistResponseSchema } from "./basePlaylistForPlaylistResponse";
import { BaseSongWithAlbumResponseSchema } from "./baseSongWithAlbumResponse";
import { BaseStationResponseSchema } from "./baseStationResponse";
import { BaseVideoResponseSchema } from "./baseVideoResponse";
import { PlaylistContributorResponseSchema } from "./playlistContributorResponse";
import { PlaylistResponseItemSchema } from "./playlistResponseItem";

export const BasePlaylistResponseSchema = z.object({
    type: z.union([z.literal("playlist")]).default("playlist"),
    description: z.string().nullable(),
    provider: z.string(),
    publicId: z.string(),
    url: z.string(),
    providerUrl: z.string(),
    name: z.string(),
    medias: z.array(
        z.union([
            z
                .lazy(() => PlaylistResponseItemSchema)
                .unwrap()
                .extend({
                    item: z.union([
                        z.lazy(() => BaseSongWithAlbumResponseSchema),
                    ]),
                }),
            z
                .lazy(() => PlaylistResponseItemSchema)
                .unwrap()
                .extend({
                    item: z.union([z.lazy(() => BaseVideoResponseSchema)]),
                }),
            z
                .lazy(() => PlaylistResponseItemSchema)
                .unwrap()
                .extend({
                    item: z.union([z.lazy(() => BaseStationResponseSchema)]),
                }),
            z
                .lazy(() => PlaylistResponseItemSchema)
                .unwrap()
                .extend({
                    item: z.union([
                        z.lazy(() => BasePlaylistForPlaylistResponseSchema),
                    ]),
                }),
            z
                .lazy(() => PlaylistResponseItemSchema)
                .unwrap()
                .extend({
                    item: z.union([
                        z.lazy(() => BaseAlbumWithoutSongsResponseSchema),
                    ]),
                }),
            z
                .lazy(() => PlaylistResponseItemSchema)
                .unwrap()
                .extend({
                    item: z.union([
                        z.lazy(() => BaseAlbumWithSongsResponseSchema),
                    ]),
                }),
        ])
    ),
    contributors: z.array(z.lazy(() => PlaylistContributorResponseSchema)),
    imageUrl: z.string(),
    owner: z.string(),
});

export type BasePlaylistResponse = z.infer<typeof BasePlaylistResponseSchema>;
